### ISSUE TITLE
Request's Body is always null using RoutingDsl 

### ISSUE THREAD

**Issue by amatosc** [playframework/playframework#7682]
### Play Version (2.5.x / etc)
2.6.2

### API (Scala / Java / Neither / Both)
Java

### Operating System (Ubuntu 15.10 / MacOS 10.10 / Windows 10)
` Versión del sistema:	macOS 10.12.6
  Versión del kernel:	Darwin 16.7.0
`

### JDK (Oracle 1.8.0_72, OpenJDK 1.8.x, Azul Zing)
```sh
java version "1.8.0_131"
Java(TM) SE Runtime Environment (build 1.8.0_131-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.131-b11, mixed mode)
````
### Expected Behavior
Request's body be defined. 
1. Do a request to any of your endpoint
2. get the body as json (`request().body()`)

### Actual Behavior
Actually the body is always `null`, but the method `hasBody` return `true`. Note that I'm using `routingDsl`. I've got a Controller, routing and app loader. I have been following this [doc](https://www.playframework.com/documentation/2.6.x/JavaRoutingDsl#Providing-a-DI-router-with-Guice) 

I'm sending a valid request (content-type=application/json and the body is a valid json.) 

```java
@Singleton
public class Routing implements Provider<play.api.routing.Router> {
    private static final String URL_LOGIN = "/login";

    private final Config config;
    private RoutingDsl routingDsl;
    private Secure secure;

    private boolean isRunningAsInstance;
    private String prefix;

    @Inject
    public Routing(RoutingDsl dsl, Config config, Secure secure) {
        this.routingDsl = dsl;
        this.config = config;
        this.secure = secure;
    }

    @Override
    public play.api.routing.Router get() {
        return routingDsl
                .POST(getUrl(URL_LOGIN)).routeTo(secure::login)
                .build()
                .asScala();
    }
}

public class AppLoader extends GuiceApplicationLoader {

    @Override
    protected GuiceableModule[] overrides(Context context) {
        GuiceableModule[] modules = super.overrides(context);
        GuiceableModule module = GuiceableModule$.MODULE$.fromPlayBinding(new BindingKey<>(play.api.routing.Router.class)
                .toProvider(Routing.class).eagerly());

        List<GuiceableModule> copyModules = new ArrayList<>(Arrays.asList(modules));
        copyModules.add(module);

        return copyModules.toArray(new GuiceableModule[copyModules.size()]);
    }
}


public class Secure extends Controller {
    private final Handler handler;

    @Inject
    public Secure(Handler handler) {
        this.handler = handler;
    }
    public Result login() {
        Logger.debug("Secure.login: starring login");
        JsonNode body = null;

        if (request().hasBody()) {
            body = request().body().asJson();
            return ok();
        }
        return badRequest();
    }
}
```

**Comment by marcospereira** [comment]
Thank you for the report, @amatosc.

This is happening because in the `RouterBuilderHelper` we are [creating the Java `Http.Context` using a `RequestHeader`](https://github.com/playframework/playframework/blob/2.6.2/framework/src/play-java/src/main/scala/play/routing/RouterBuilderHelper.scala#L66) and that eventually creates a Java `Request` using a null body:

https://github.com/playframework/playframework/blob/2.6.2/framework/src/play/src/main/java/play/mvc/Http.java#L907-L909

`hasBody` [checks the `Content-Length` header](https://github.com/playframework/playframework/blob/2.6.2/framework/src/play/src/main/scala/play/api/mvc/Headers.scala#L42-L45) and returns true since it is there, but the body itself was discarded before, then the error you are seeing.

I will work on a fix for this problem.

**Comment by amatosc** [comment]
Hi @marcospereira, thanks very much for your change.

Is there any way to fix this, until this change is released?  I tried create my own instance of `RoutingDsl` in instance of inject it, but I got the same result.

PS: Sorry for my dummy question, but I'm a bit new using play 2

**Comment by stefanhartmann123** [comment]
Is there any fix how to solve this before the new version is released?

**Comment by richdougherty** [comment]
@stefanhartmann123, yes it was fixed by #7684 and #7696.

**Comment by richdougherty** [comment]
Hi @amatosc and @stefanhartmann123, we've just released Play 2.6.3 (https://blog.playframework.com/play-2-6-3-released/). If you get a chance it would be great if you could check that this issue is fixed for you. :)

**Comment by richdougherty** [comment]
The fix was made to core Play module, so if you update Play to version 2.6.3 that's all you should need to do. You can verify you're using the correct version by typing `show dependencyClasspath` inside `sbt` and checking that the `play-java` version is `2.6.3`.

I'll re-open this ticket again for now. FYI, here is the unit test that verifies that request bodies are non-null. Perhaps we didn't capture your situation properly in this test:

https://github.com/playframework/playframework/pull/7696/files#diff-a281ba3461ae2127ccba3ae3a8b2d67cR36

**Comment by amatosc** [comment]
This the output of the command, I checked it and all Play's packages that I found are from 2.6.3 version. 

```sbt
$ show dependencyClasspath
 [info] * Attributed(~/.ivy2/cache/com.typesafe.play/play-java_2.12/jars/play-java_2.12-2.6.3.jar)
```

I saw the test and I think, that maybe there are something diferente between how your `application()` method created the app and my loader class, but as I said I just followed the doc.

**Comment by richdougherty** [comment]
Thanks for checking. I'll leave this open because we'll need to try and reproduce this bug again.

**Comment by marcospereira** [comment]
Thanks for checking, @amatosc.

This is now another problem. Before, we were completely losing the body and now we are not able to use `asJson` because the body is not a `JsonNode`, but instead an `AnyContentAsJson`. That is happening because `RoutingDsl` uses a Scala body parser, not a Java one. This could be annoying to fix in a backward compatible way.

I would say that the ideal way is to use a Java body parser just like we do for regular actions. Anyway, I will dig a little more to see how is the best way to fix this for 2.6.x branch.

Again, thanks for the help, @amatosc. :-)

**Comment by luca-emme** [comment]
Hi @marcospereira any news about this issue? Is there any workaround to read json body with RoutingDsl?
Thank you!

**Comment by marcospereira** [comment]
Hi @luca-emme, we still don't have a solution for this problem.

As explained in my previous comment above, the problem here is that, since `RoutingDsl` is using a Scala body parser, we are getting a `play.api.mvc.AnyContentAsJson` instead of a `com.fasterxml.jackson.databind.JsonNode`. I don't think we currently have a quick workaround.

But keep in mind we are talking about having this prioritized over the next couple weeks. Of course, we appreciate if someone could make a contribution here.

**Comment by marcospereira** [comment]
Hi @amatosc, @stefanhartmann123, and @luca-emme,

This issue was fixed in 2.6.9. Could you please check and confirm it is working for you guys?

**Comment by luca-emme** [comment]
Hi @marcospereira I confirm that version 2.6.9 works as expected. 
Thank you!
